### PR TITLE
Implement a ZeroMQ connector

### PR DESCRIPTION
**Pull request by mavam**
This PR adds a [ZeroMQ](https://zeromq.org/) connector with the following socket types:

- Loader: `SUB`
- Saver: `PUB`

The assumption is that the pipeline sink *publishes* messages and that the pipeline source *subscribes* to messages. Independent of that is the socket direction: you can use `--bind` or `--connect` to control the socket establishment. By default, the sink binds as publisher and source connects as subscriber.

### CODE REVIEW

File: .github/workflows/tenzir.yaml
Comment (dominiklohmann) [inline .github/workflows/tenzir.yaml:856]: Nit: both of the lists here used to be alphabetically sorted.

File: plugins/zmq/src/plugin.cpp
Comment (dominiklohmann) [inline plugins/zmq/src/plugin.cpp:398]: You add the positional argument to the parser twice here.

File: web/blog/tenzir-v4.2/index.md
Comment (dominiklohmann) [inline web/blog/tenzir-v4.2/index.md:5]: `draft: true`

File: plugins/zmq/src/plugin.cpp
Comment (dominiklohmann) [inline plugins/zmq/src/plugin.cpp:207]: Can you add a comment here that explains what -1ms indicates?

**Review by dominiklohmann — COMMENTED** [review]
I played with this and can confirm it works well in practice. We'll have to figure out the fix for detached operators sometime soon, though, and this still needs a more thorough review than me scrolling through the code for 20 minutes.

File: plugins/zmq/src/plugin.cpp
Comment (mavam) [reply plugins/zmq/src/plugin.cpp:398]: Fixed in d7b5490d85.

File: .github/workflows/tenzir.yaml
Comment (mavam) [reply .github/workflows/tenzir.yaml:856]: 34a3e812af

File: plugins/zmq/src/plugin.cpp
Comment (mavam) [reply plugins/zmq/src/plugin.cpp]: 45e726759e

File: web/blog/tenzir-v4.2/index.md
Comment (Dakostu) [inline web/blog/tenzir-v4.2/index.md:19]: I should probably do this. 
After this PR has been merged - then I add my S3 stuff to this file in my PR? @mavam

Comment (mavam) [reply web/blog/tenzir-v4.2/index.md:19]: Exactly what I was thinking.

File: plugins/zmq/src/plugin.cpp
Comment (Dakostu) [inline plugins/zmq/src/plugin.cpp:311]: I "tapped into the raw feed" of my `nic` pipeline, as per the blog post, to receive data, and my `load zmq` pipeline always finishes after receiving the first message. Is this the tapping behavior that you'd expect?

I changed this conditional to a `while` loop, and I got the full-on continuous tap with `nic`. But transmitting non-stream data, e.g. via `from file`, won't terminate the reading pipeline anymore. 

Tired train ride brainstorming: Is there a "terminating message" that the `zmq_saver` can send and the `zmq_loader` only terminates upon receiving that message? Maybe the `engine` can fulfill such behavior?

NOTE: We should probably think about this if a non-Tenzir user wants that kind of feature.

**Review by Dakostu — APPROVED** [review]
I played around with this a bit, and I approve. This can be :ship: ped.

I did leave some comments though.

File: web/docs/connectors/zmq.md
Comment (mavam) [reply web/docs/connectors/zmq.md]: Good point. I reworded the example in 84c338d171 to make it clearer what's going on.

File: web/blog/tenzir-v4.2/index.md
Comment (mavam) [reply web/blog/tenzir-v4.2/index.md]: 204ae259d6

File: plugins/zmq/src/plugin.cpp
Comment (mavam) [reply plugins/zmq/src/plugin.cpp:311]: > I "tapped into the raw feed" of my nic pipeline, as per the blog post, to receive data, and my load zmq pipeline always finishes after receiving the first message. Is this the tapping behavior that you'd expect?

No, that's not what I'd expect. I'm glad you tested this. I can repro it and will work on a fix.

When I tested this in more depth, I also realized that a disconnected 0mq socket doesn't stop the source from producing after it started once. This needs more accurate connection management.

> I changed this conditional to a `while` loop, and I got the full-on continuous tap with `nic`. But transmitting non-stream data, e.g. via `from file`, won't terminate the reading pipeline anymore.
>
> Tired train ride brainstorming: Is there a "terminating message" that the `zmq_saver` can send and the `zmq_loader` only terminates upon receiving that message? Maybe the `engine` can fulfill such behavior?

This idea goes in the right direction. The proper fix would be processing the 0mq meta events from monitor. There are basically two socket, the monitoring socket and the data socket. If we get an "disconnect" event from the monitoring socket, we should suspend execution.

A `while` loop can't work, but we need the equivalent of a "main loop" that checks both channels. I'll take a stab at it.

Comment (mavam) [reply plugins/zmq/src/plugin.cpp]: @Dakostu I was able to fix the issue by fixing the main loop equivalent in saver/loader land. Mind trying it again?

If it works now, just resolve the issue; I've enabled auto-merge.

Comment (Dakostu) [reply plugins/zmq/src/plugin.cpp]: OK, so now I can tap into `nic` data, but receiving pipelines never terminate, which also includes finite data `from file` delivered via `zmq`.

The `zmq_loader`'s `make` loop does not seem to handle the case where this "disconnect" event happens - the `num_peers` does get decreased in the `poll_monitor`, but with no `num_peers`, the generator is just stuck `co_yield`ing empty chunks. We also can't just `co_return`, it would make receiving pipelines non-deterministic; it already runs into that case while establishing connection.

You probably need a second counter `num_disconnected_peers` that increases with each disconnect event - then in the `make` loop, if `num_peers` is 0, check if `num_disconnected_peers` is not 0, then `co_return`, else `co_yield` empty chunks. Good idea?

Comment (Dakostu) [reply plugins/zmq/src/plugin.cpp]: I quickly implemented that idea and it looks like it works now. Should I push it?

Comment (mavam) [reply plugins/zmq/src/plugin.cpp]: I would argue that receiving pipelines should never terminate, or at least not by default. The reason is pub/sub semantics: publishers and subscribes come and go. What's stable is the channel in the middle.

I think you're proposing a flag that's similar to `kafka`'s `-e` to exit, right?

Comment (mavam) [reply plugins/zmq/src/plugin.cpp]: We discussed this issue in a all and agreed that we are going to leave the issue untouched for this PR, i.e., ship as-is.

The reasoning is that we'd be making strong use case assumptions to terminate a subscriber when the peer count drops to 0, which may work for some but break other use cases. We'd rather wait for user feedback before preemptively adding a `-e` flag.

**Pull request merged**
